src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs

```diff
@@ -31,24 +37,160 @@ internal ModelOperationsCatalog(IHostEnvironment env)
+        {
+            _env.CheckValue(stream, nameof(stream));
+
+            using (var rep = RepositoryReader.Open(stream, _env))
+            {
+                var entry = rep.OpenEntryOrNull(SchemaEntryName);
+                if (entry != null)
+                {
+                    var loader = new BinaryLoader(_env, new BinaryLoader.Arguments(), entry.Stream);
+                    inputSchema = loader.Schema;
+                    ModelLoadContext.LoadModel<ITransformer, SignatureLoadModel>(_env, out var transformerChain, rep,
+                        CompositeDataLoader<object, ITransformer>.TransformerDirectory);
+                    return transformerChain;
+                }
+
+                ModelLoadContext.LoadModelOrNull<IDataLoader<IMultiStreamSource>, SignatureLoadModel>(_env, out var dataLoader, rep, null);
+                if (dataLoader == null)
+                {
+                    // Try to see if the model was saved without a loader or a schema.
+                    if (ModelLoadContext.LoadModelOrNull<ITransformer, SignatureLoadModel>(_env, out var transformerChain, rep,
+                        CompositeDataLoader<object, ITransformer>.TransformerDirectory))
+                    {
+                        inputSchema = null;
+                        return transformerChain;
+                    }
+
+                    // Try to load from legacy model format.
+                    try
+                    {
+                        var loader = ModelFileUtils.LoadLoader(_env, rep, new MultiFileSource(null), false);
+                        inputSchema = loader.Schema;
+                        return TransformerChain.LoadFromLegacy(_env, stream);
+                    }
+                    catch (Exception ex)
+                    {
+                        throw _env.Except(ex, "Could not load legacy format model");
+                    }
+                }
+                if (dataLoader is CompositeDataLoader<IMultiStreamSource, ITransformer> composite)
+                {
+                    inputSchema = composite.Loader.GetOutputSchema();
+                    return composite.Transformer;
+                }
+                inputSchema = dataLoader.GetOutputSchema();
+                return new TransformerChain<ITransformer>();
+            }
+        }
 
         /// <summary>
-        /// Load the model from a file path.
+        /// Load the model and its input schema from the stream.
         /// </summary>
-        /// <param name="modelPath">Path to model.</param>
-        /// <returns>The loaded model.</returns>
-        public ITransformer Load(string modelPath)
+        /// <param name="stream">A readable, seekable stream to load from.</param>
+        /// <returns>A model of type <see cref="CompositeDataLoader{IMultiStreamSource, ITransformer}"/> containing the loader
+        /// and the transformer chain.</returns>
+        public IDataLoader<IMultiStreamSource> Load(Stream stream)
```

Comment:
>IDataLoader [](start = 15, length = 11)

it's kinda sad what now we have `IDataLoader` along side with `ITransformer` since it's increase complexity. Would be nice to have some documentation/samples which would explain in which case I need to save/load one thing, and in which another.

Also can I now have slot names in my model?